cunumeric/deferred.py

```diff
@@ -305,37 +305,250 @@ def get_scalar_array(self):
+        # N is number of index arrays
+        # shape of the output array should be the same as the shape of each
+        # index array
+        # NOTE: We need to instantiate a RegionField of non-primitive
+        # dtype, to store N-dimensional index points, to be used as the
+        # indirection field in a copy.
+        # Such dtypes are technically not supported,
+        # but it should be safe to directly create a DeferredArray
+        # of that dtype, so long as we don't try to convert it to a
+        # NumPy array.
+        N = self.ndim
+        pointN_dtype = self.runtime.get_point_type(N)
+        store = self.context.create_store(
+            pointN_dtype, shape=out_shape, optimize_scalar=True
+        )
+        output_arr = DeferredArray(
+            self.runtime, base=store, dtype=pointN_dtype
+        )
+
+        # call ZIP function to combine index arrays into a singe array
+        task = self.context.create_task(CuNumericOpCode.ZIP)
+        task.add_output(output_arr.base)
+        task.add_scalar_arg(self.ndim, ty.int64)  # N of points in Point<N>
+        task.add_scalar_arg(key_dim, ty.int64)  # key_dim
+        task.add_scalar_arg(start_index, ty.int64)  # start_index
+        for a in arrays:
+            task.add_input(a)
+            task.add_alignment(output_arr.base, a)
+        task.execute()
+
+        return output_arr
+
+    def _copy_store(self, store):
+        store_to_copy = DeferredArray(
+            self.runtime,
+            base=store,
+            dtype=self.dtype,
+        )
+        store_copy = self.runtime.create_empty_thunk(
+            store_to_copy.shape,
+            self.dtype,
+            inputs=[store_to_copy],
+        )
+        store_copy.copy(store_to_copy, deep=True)
+        return store_copy, store_copy.base
+
+    def _create_indexing_array(self, key, is_set=False):
+        store = self.base
+        rhs = self
+        # the index where the first index_array is passed to the [] operator
+        start_index = -1
+        if (
+            isinstance(key, NumPyThunk)
+            and key.dtype == np.bool
+            and key.shape == rhs.shape
+        ):
+            if not isinstance(key, DeferredArray):
+                key = self.runtime.to_deferred_array(key)
+
+            out_dtype = rhs.dtype
+            # in cease this operation is called for the set_item, we
+            # return Point<N> type field that is later used for
+            # indirect copy operation
+            if is_set:
+                N = rhs.ndim
+                out_dtype = rhs.runtime.get_point_type(N)
+
+            out = rhs.runtime.create_unbound_thunk(out_dtype)
+            task = rhs.context.create_task(CuNumericOpCode.ADVANCED_INDEXING)
+            task.add_output(out.base)
+            task.add_input(rhs.base)
+            task.add_input(key.base)
+            task.add_scalar_arg(is_set, bool)
+            task.add_alignment(rhs.base, key.base)
+            task.execute()
```

Comment:
You still need to do `add_broadcast([all inputs], axes=[all except 0])` in this case, to make sure that the final result, after the outputs from all point tasks are coallesced, is in global row-major order.

For example, consider the following 2x3 boolean array, that has been split in two tiles on dimension 1:

```
False  False | True
False  True  | False
```

The first point task, that gets the left tile, will produce the output index `(1, 1)`, while the second point task will produce `(0,2)`. 

The output sub-regions will be combined in the same order as the tiles, giving us `[(1,1), (0,2)]`, which is not in row-major order.


Conversely, if we demand that only dimension 0 be partitioned, then legate will have to split by rows:

```
False  False True
-----------------
False True  False
```

and the final output will be in global row-major order.

This will remain true no matter how large the rows are, since any element in output sub-region `i` will be smaller on dimension 0 than all elements in output sub-region `i+1` (and assuming the kernels make sure to internally list indices in row-major order).

Reply:
Yes, thank you!